### PR TITLE
[Bugfix] Fix Maverick correctness by filling zero to cache space in cutlass_moe

### PR DESCRIPTION
**Pull request by minosfuture**
## Purpose

#19667 changed the workspace creation from `torch.zeros` to `torch.empty`. This ends up causing correctness for models using cutlass_moe, e.g. Maverick in our test case. This PR fixes the correctness issue by explicitly filling zeros in cutlass_moe.

## Test Plan

* lm_eval
* added ut that would fail without this fix

## Test Result

lm_eval results:

local-chat-completions (model=meta-llama/Llama-4-Maverick-17B-128E-Instruct-FP8,base_url=http://127.0.0.1:8081/v1/chat/completions,num_concurrent=32), gen_kwargs: (None), limit: 200.0, num_fewshot: 5, batch_size: 1
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value|   |Stderr|
|-----|------:|----------------|-----:|-----------|---|----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.935|±  |0.0175|
|     |       |strict-match    |     5|exact_match|↑  |0.920|±  |0.0192|

unit test stability verified:
* without `c1.fill_(0)`, the following one liner verifies stable failure:
```bash
for i in {1..10}; do echo $i; pytest -s tests/kernels/moe/test_cutlass_moe.py  -k "test_run_cutlass_moe_fp8 or test_cutlass_moe_8_bit_EP_large" -v  2>&1 > /dev/null && { echo "shouldn't succeed"; exit 1; } done
```
* with `c1.fill_(0)`, the following verifies stable success:
```bash
for i in {1..10}; do echo $i; pytest -s tests/kernels/moe/test_cutlass_moe.py  -k "test_run_cutlass_moe_fp8 or test_cutlass_moe_8_bit_EP_large" -v  2>&1 > /dev/null || { echo "should succeed"; exit 1; } done
```

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py [hunk 1]
```diff
@@ -176,6 +176,7 @@ def run_cutlass_moe_fp8(
         c1 = _resize_cache(workspace13, (M * topk, N * 2))
         c2 = _resize_cache(workspace2, (M * topk, N))
         c3 = _resize_cache(workspace13, (M * topk, K))
+        c1.fill_(0)
```
Comment (yeqcharlotte) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: sounds like this should impact both chunking and non chunking path?

File: tests/kernels/moe/test_cutlass_moe.py [hunk 2]
```diff
@@ -365,3 +369,131 @@ def test_cutlass_moe_8_bit_EP(
                                    cutlass_output,
                                    atol=5e-2,
                                    rtol=1e-2)
+
+
+@pytest.mark.parametrize("m,n,k,topk", [(1, 8192, 5120, 31)])
```
Comment (yeqcharlotte) [inline tests/kernels/moe/test_cutlass_moe.py]: let's have some m > 32k

File: tests/kernels/moe/test_cutlass_moe.py [hunk 3]
```diff
@@ -1,16 +1,20 @@
...
-from vllm.model_executor.layers.fused_moe.cutlass_moe import cutlass_moe_fp8
+from vllm.model_executor.layers.fused_moe.cutlass_moe import (
+    cutlass_moe_fp8, run_cutlass_moe_fp8)
 from vllm.model_executor.layers.fused_moe.fused_moe import (fused_experts,
                                                             fused_topk)
+from vllm.model_executor.layers.fused_moe.utils import (
+    moe_kernel_quantize_input)
 from vllm.platforms import current_platform
 
 NUM_EXPERTS = [40, 64]
```
Comment (yeqcharlotte) [inline tests/kernels/moe/test_cutlass_moe.py:20]: does it help with the working sets at line 38-39?

**Review by yeqcharlotte — COMMENTED** [review]
thanks for the unit test to repro the issue!

cc: @bnellnm to also take a look!

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 1 above)
Comment (ElizaWszola) [inline vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: Is this needed when we don't use `expert_map`? In case it's not, can you write a condition for this?

File: tests/kernels/moe/test_cutlass_moe.py [hunk 4]
```diff
@@ -365,3 +369,131 @@ def test_cutlass_moe_8_bit_EP(
...
+    m: int,
+    n: int,
+    k: int,
+    e: int,
+    topk: int,
+    per_act_token: bool,
+    per_out_channel: bool,
+    ep_size: int,
+):
+    current_platform.seed_everything(7)
```
Comment (bnellnm) [inline tests/kernels/moe/test_cutlass_moe.py]: If the body of this is the same as `test_cutlass_moe_8_bit_EP` can you factor it out into a common function?

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 1 above)
Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: I think the non-chunking (batched) path is not impacted because c1 is fully overridden. I don't have solid proof though (I need to look into that code path more). @bnellnm / @ElizaWszola comments?

File: tests/kernels/moe/test_cutlass_moe.py (hunk 2 above)
Comment (minosfuture) [reply tests/kernels/moe/test_cutlass_moe.py]: added

File: tests/kernels/moe/test_cutlass_moe.py (hunk 3 above)
Comment (minosfuture) [reply tests/kernels/moe/test_cutlass_moe.py:20]: unfortunately no. We can look into this more separately.

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 1 above)
Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: thanks. updated!

File: tests/kernels/moe/test_cutlass_moe.py (hunk 4 above)
Comment (minosfuture) [reply tests/kernels/moe/test_cutlass_moe.py]: Good point! thx. Updated with refactoring a few similar test functions here.

File: tests/kernels/moe/test_cutlass_moe.py [hunk 5]
```diff
@@ -337,9 +348,62 @@ def test_cutlass_moe_8_bit_EP(
...
+    n: int,
+    k: int,
+    e: int,
+    topk: int,
+    per_act_token: bool,
+    per_out_channel: bool,
+    ep_size: int,
 ):
     current_platform.seed_everything(7)
-    monkeypatch.setenv("VLLM_FUSED_MOE_CHUNK_SIZE", "8192")
```
Comment (minosfuture) [inline tests/kernels/moe/test_cutlass_moe.py:345]: this is not very git-friendly, but note this line is not removed during refactoring. see `test_cutlass_moe_8_bit_no_graph`

File: tests/kernels/moe/test_cutlass_moe.py [hunk 6]
```diff
@@ -365,3 +369,129 @@ def test_cutlass_moe_8_bit_EP(
...
+        workspace13.random_()
+        output_random_workspace = torch.empty(output_shape,
+                                              device="cuda",
+                                              dtype=mt.a.dtype)
+        func(output_random_workspace)
+
+        workspace13.fill_(0)
+        output_zero_workspace = torch.zeros(output_shape,
+                                            device="cuda",
+                                            dtype=mt.a.dtype)
```
[the remark this answers is not in the document]
Comment (minosfuture) [reply tests/kernels/moe/test_cutlass_moe.py:462]: invalid suggestion. ignore

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 1 above)
Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: I tested this locally and found that the batched case needs to be cleared also.  I think it's probably best to unconditionally zero out `c1`

Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: I see. updated.  Could you share how to run batched case tests? thx.

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: I ran `test_cutlass_moe.py` and `test_pplx_cutlass_moe.py`

Comment (bnellnm) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: I think the. condition should be `expert_map is not None or self.use_batched_format`.   Batched mode is almost always going to have some garbage space in the tensor.

Comment (tlrmchlsmth) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: Why does the expert map matter here?

Comment (tlrmchlsmth) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: Is the failure caused by reading in garbage data for dynamic per-tensor quantization?

Does it work in the static per-tensor case?

**Review by tlrmchlsmth — APPROVED** [review]
A couple of questions but good to land once rebased. Thanks for the fix

File: vllm/model_executor/layers/fused_moe/cutlass_moe.py (hunk 1 above)
Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: yep, exactly. And it works, without zero-out, in the static scale case.

Comment (minosfuture) [reply vllm/model_executor/layers/fused_moe/cutlass_moe.py:182]: only when expert_map is in use, the c1 will not be fully overridden. Note that c1 is resized into (# token globally, 2N), while the actually used space is (# token locally, 2N). When expert_map is in use, the local token size can be smaller.

**Comment by minosfuture** [comment]
@tlrmchlsmth this should be good to go. Could you help trigger CI/auto-merge? thx!

**Review by tlrmchlsmth — APPROVED** [review]
